File: src/main/java/org/gga/graph/connection/StrongComponents.java
package org.gga.graph.connection;

import com.google.common.base.Preconditions;
import org.gga.graph.Edge;
import org.gga.graph.Graph;
import org.gga.graph.search.dfs.AbstractDfsVisitor;
import org.gga.graph.util.IntStack;

/**
 * @author mike
 */
public class StrongComponents {

    public static int strongComponents(
            Graph graph,
            int[] componentMap,
            int[] rootMap) {
        Preconditions.checkArgument(graph.isDirected());

        TarjanSccVisitor visitor = new TarjanSccVisitor(rootMap, componentMap, new int[graph.V()]);
        graph.getDfs().depthFirstSearch(visitor);

        return visitor.currentComponent;
    }

    public static boolean isConnected(Graph graph) {
        final int[] starts = {0};
        graph.getDfs().depthFirstSearch(new AbstractDfsVisitor() {
            @Override
            public void startVertex(int v, Graph graph) {
                starts[0]++;
            }
        });

        return starts[0] == 1;
    }


    private static class TarjanSccVisitor extends AbstractDfsVisitor {
        private final int[] rootMap;
        private final int[] componentMap;
        private final int[] discoverTime;
        private int dfsTime = 0;
        private int currentComponent = 0;
        private final IntStack stack = new IntStack();

        private TarjanSccVisitor(int[] rootMap, int[] componentMap, int[] discoverTime) {
            this.rootMap = rootMap;
            this.componentMap = componentMap;
            this.discoverTime = discoverTime;
        }

        @Override
        public void discoverVertex(int v, Graph graph) {
            rootMap[v] = v;
            componentMap[v] = Integer.MAX_VALUE;
            discoverTime[v] = dfsTime++;
            stack.push(v);
        }

        @Override
        public void finishVertex(int v, Graph graph) {
            for (Edge e : graph.getEdges(v)) {
                int w = e.other(v);

                if (componentMap[w] == Integer.MAX_VALUE) {
                    rootMap[v] = minDiscoverTime(rootMap[v], rootMap[w]);
                }
            }

            if (rootMap[v] == v) {
                int w;

                do {
                    w = stack.pop();
                    componentMap[w] = currentComponent;
                    rootMap[w] = v;
                } while (w != v);

                currentComponent++;
            }
        }

        private int minDiscoverTime(int v, int w) {
            return discoverTime[v] < discoverTime[w] ? v : w;
        }
    }
}

File: src/main/scala/org/gga/graph/connection/StrongComponents.scala
package org.gga.graph.connection

import com.google.common.base.Preconditions
import org.gga.graph.Graph
import org.gga.graph.search.dfs.AbstractDfsVisitor
import collection.mutable

/**
 * @author mike
 */
object StrongComponents {
  def strongComponents(graph: Graph, componentMap: Array[Int], rootMap: Array[Int]): Int = {
    Preconditions.checkArgument(graph.isDirected)
    val visitor = new StrongComponents.TarjanSccVisitor(rootMap, componentMap, new Array[Int](graph.V))
    graph.getDfs.depthFirstSearch(visitor)
    visitor.currentComponent
  }

  private class TarjanSccVisitor(aRootMap: Array[Int], aComponentMap: Array[Int], aDiscoverTime: Array[Int]) extends AbstractDfsVisitor {
    private val rootMap = aRootMap
    private val componentMap = aComponentMap
    private val discoverTime = aDiscoverTime

    private var dfsTime: Int = 0
    var currentComponent: Int = 0
    private final val stack: mutable.Stack[Int] = new mutable.Stack

    override def discoverVertex(v: Int, graph: Graph) {
      rootMap(v) = v
      componentMap(v) = Integer.MAX_VALUE
      discoverTime(v) = ({
        dfsTime += 1; dfsTime
      })
      stack.push(v)
    }

    override def finishVertex(v: Int, graph: Graph) {
      for (e <- graph.edges(v)) {
        val w: Int = e.other(v)
        if (componentMap(w) == Integer.MAX_VALUE) {
          rootMap(v) = minDiscoverTime(rootMap(v), rootMap(w))
        }
      }
      if (rootMap(v) == v) {
        var w: Int = 0
        do {
          w = stack.pop()
          componentMap(w) = currentComponent
          rootMap(w) = v
        } while (w != v)
        currentComponent += 1
      }
    }

    private def minDiscoverTime(v: Int, w: Int): Int = {
      if (discoverTime(v) < discoverTime(w)) v else w
    }
  }
}